### DISCUSSION TITLE
Feature research / roadmap visibility

### DISCUSSION THREAD

**Discussion by JesseCoretta** [389ds/389-ds-base#6384] [Ideas] [answered]
This is a discussion meant for the senior/lead developers of 389DS.    I have reached out numerous times to your IRC channel, no one responds, hence I am going to reach out here.

I am a directory services SME.  I am interested to learn whether there are any implementation plans related to the following capabilities:

  - DIT structure rule governance w/ name forms (see [ITU-T Rec. X.501](https://www.itu.int/rec/t-rec-x.501/en))
  - Collective attributes w/ subtreeSpecification ([RFC3671](https://www.rfc-editor.org/rfc/rfc3671.html) and [RFC3672](https://www.rfc-editor.org/rfc/rfc3672.html))
  - Transactional support through [RFC 5805](https://datatracker.ietf.org/doc/html/rfc5805)

Are _any_ of these planned on the roadmap? I am offering my skills in the area of research related to an effective implementation, but I wanted to reach out before doing any of this.

**Comment by vashirov** [comment]
Hi @JesseCoretta, thanks for reaching out.

>  I have reached out numerous times to your IRC channel, no one responds, hence I am going to reach out here.

Due to the synchronous nature of IRC, if you won't stay online on the channel, nobody can reply to you. I saw your messages, but you were offline at that time.
Starting discussion here or on [389-users mailing list](https://lists.fedoraproject.org/archives/list/389-users@lists.fedoraproject.org/) is a good idea, since our team works in different time zones.

> Are any of these planned on the roadmap?

Feature development is driven by customers' demand and their business needs. At the moment I'm not aware of any requests for the features you listed. All RFEs that are potentially on our roadmap can be found [here](https://github.com/389ds/389-ds-base/issues?q=is%3Aopen+is%3Aissue+label%3ARFE+) on GitHub or in RHDS Jira issue [tracker](https://issues.redhat.com/issues/?jql=project%20%3D%20DIRSRV%20AND%20issuetype%20%3D%20Story%20AND%20resolution%20%3D%20Unresolved%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC).

If you're interested in the capabilities you listed, please open a [Feature Request](https://github.com/389ds/389-ds-base/issues/new?assignees=&labels=needs+triage&projects=&template=feature_request.md&title=) for each of them. We need to evaluate the requests first.

**Comment by Firstyear** [comment, marked as answer]
The question that should be asked isn't if we plan to implement them - it's if anyone will actually *use* them.

As a higher level commentary standards like RFC's have two forms - the whole standard as written, and the subset that is used. This subset will be what makes up 99% of the use of the service. The remainder is for all intents and purposes, useless.

For example when we look at the collection of LDAP standards, there exists a huge surface area - but it doesn't take long to find the parts that are never used within LDAP. 

So when we consider these suggestions, are we implement the "standard for the standards sake"? Or are we implementing something where there is actually a need for it to be used, and used by a not-insignificant number of users?

> 
>     * DIT structure rule governance w/ name forms (see [ITU-T Rec. X.501](https://www.itu.int/rec/t-rec-x.501/en))

Given the direction of identity services and the industry as a whole, X.501 is not really relevant to business interests. I don't think this is worth pursuing unless there is a really compelling use case. 

> 
>     * Collective attributes w/ subtreeSpecification ([RFC3671](https://www.rfc-editor.org/rfc/rfc3671.html) and [RFC3672](https://www.rfc-editor.org/rfc/rfc3672.html))

While both of these seem interesting, they are also very complex to what they need to achieve. 

Additionally, given both of these were created in 2003 and we haven't had any requests for them until now in 2021, I would safely assign these to the bucket of "no interest" from directory operators.

> 
>     * Transactional support through [RFC 5805](https://datatracker.ietf.org/doc/html/rfc5805)

This would be the only extension I think that is actually interesting to implement since it effectively is a "batch update" like https://www.rfc-editor.org/rfc/rfc7644#section-3.7 . The actual issue here is that we don't have the infrastructure within 389-ds to handle this properly, and would pretty much necessitate a complete rewrite of transaction handling in the server to achieve.

To explain more, in other systems you would expect a DB to have ACID properties. 389-ds doesn't have Atomicity or Isolation. We only have Consistency and Durability. 

This is because within a "modify" small writes are performed to entries and other related entries that can be interleaved with other simultaneous operations. You can't "undo" a modify and expect it to undo the changes that happened to refint or memberof etc. 

OpenLDAP and Kanidm do have this because they implement COW and have true ACID behaviour, where the whole transaction can be rolled backed. 

As a result 389-ds won't be able to have this rfc implemented *unless* the whole transaction model of the server is changed to have read/write txn's taken at a higher level and implementing true COW behaviour to match LMDB's behaviour.

**Comment by JesseCoretta** [comment]
thanks for responding
